include/picongpu/fields/background/templates/twtstight/BField.tpp

```diff
@@ -26,174 +26,147 @@
 #include <pmacc/math/Complex.hpp>
 #include <pmacc/types.hpp>
 
-#include <tuple>
-
-namespace picongpu
+namespace picongpu::templates::twtstight
 {
-    /** Load pre-defined background field */
-    namespace templates
+    using BField = TWTSTight<FieldB>;
+
+    /** Calculate the Bx(r,t) field
+     *
+     * @param pos Spatial position of the target field.
+     * @param time Absolute time (SI, including all offsets and transformations)
+     *             for calculating the field */
+    template<>
+    HDINLINE float_T BField::calcTWTSFieldX(float3_64 const& pos, float_64 const time) const
+    {
+        auto const minimalCoordinates = defineMinimalCoordinates(pos, time);
+
+        /* To avoid underflows in computation, fields are set to zero
+         * before and after the respective TWTS pulse envelope.
+         */
+        if(isOutsideTWTSEnvelope(minimalCoordinates))
+            return float_T(0.0);
+
+        auto const commonHelperVariables = defineCommonHelperVariables(minimalCoordinates);
+
+        auto const [absPhi, sinPhi, cosPhi, beta0, tanAlpha, cspeed, lambda0, omega0, tauG, w0, k]
+            = basicTWTSHelperVariables;
+        auto const [x, y, z, t] = minimalCoordinates;
+        auto const [tanPhi, cotPhi, sinPhi_2, cosPhi_2, sinPolAngle, cosPolAngle, sin2Phi] = trigonometryShortcuts;
+        auto const [floatHelpers, complexHelpers] = commonHelperVariables;
+        auto const [x2, tauG2, psi0, w02, beta02, nu, xi, besselI0const] = floatHelpers;
+        auto const [Xm, rhom, Xm2, rhom2, besselJ0const, besselJ1const] = complexHelpers;
+
+        auto const zeroOrder = defineTWTSEnvelope(minimalCoordinates, commonHelperVariables);
+
+        complex_T const result
+            = (complex_T(0, -0.25) * math::exp(I * (omega0 * t - k * y * cosPhi)) * zeroOrder
+               * (k * rhom * besselJ0const
+                      * (cosPolAngle * (-rhom2 + x2 + x * cosPhi * Xm) * sinPhi_2
+                         - sinPolAngle * (rhom2 + rhom2 * cosPhi_2 - x2 * sinPhi_2 + x * cosPhi * sinPhi_2 * Xm))
+                  + besselJ1const
+                      * (cosPolAngle * sinPhi
+                             * (rhom2 - float_T(2.0) * x2 + I * Xm * rhom2 * (k * sinPhi)
+                                + x * cosPhi * (float_T(-2.0) * Xm - I * rhom2 * (k * sinPhi)))
+                         + sinPolAngle
+                             * ((rhom2 - float_T(2.0) * x2) * sinPhi + I * rhom2 * (-Xm + x * cosPhi) * (k * sinPhi_2)
+                                + x * sin2Phi * Xm)))
+               * psi0)
+            / (cspeed * besselI0const * rhom * rhom2);
+
+        /* A 180deg-rotation of the field vector around the y-axis
+         * leads to a sign flip in the x- and z- components, respectively.
+         * This is implemented by multiplying the result by "phiPositive".
+         */
+        return phiPositive * result.real() / sim.unit.speed();
```

Comment:
Just to make sure: The return lines of these functions look slightly different each time. (With/without `/ sim.unit.speed()` and `phiPositive`.) Would you please check that that's on purpose and not by oversight?

Reply:
I just re-checked. All x- and -z components have the `phiPositive` and all B-field components have the `/ sim.unit.speed()`. Everything is on purpose and fine.

Reply:
Thanks!